src/chunk.c

```diff
@@ -64,6 +64,95 @@ ChunkResult Uncompressed_AddSample(Chunk_t *chunk, Sample *sample) {
     return CR_OK;
 }
 
+static ChunkResult operateOccupiedSample(AddCtx *aCtx, size_t idx) {
+    Chunk *regChunk = (Chunk *)aCtx->inChunk;
+    short numSamples = regChunk->num_samples;
+    // printf("cur %lu vs sample %lu, %f\n", ChunkGetSample(regChunk, i)->timestamp, sample->timestamp, sample->value);
+    switch (aCtx->type) {
+        case UPSERT_NOT_ADD:    
+            return CR_OCCUPIED;
+        case UPSERT_ADD:        
+            regChunk->samples[idx] = aCtx->sample;
+            return CR_OK;
+        case UPSERT_DEL: { 
+            memmove(&regChunk->samples[idx],
+                    &regChunk->samples[idx + 1],
+                    (numSamples - idx) * sizeof(Sample));
+            if (numSamples == regChunk->max_samples) {
+                regChunk->num_samples = regChunk->max_samples = numSamples - 1;
+                regChunk->samples = realloc(regChunk->samples, regChunk->max_samples * sizeof(Sample));
+            }
```

Comment:
why realloc? you already shifted the data, just adjust `numSamples` and zero out the last sample for safety.

Reply:
This realloc is very cheap since it will never actually move the data.
We should decide whether we free the space or save it for future back-filling.

Reply:
there's no need to realloc, its not cheap as you don't always know whats the underline implementation, it might move data or it might not move data.
please keep the memory optimization to a later time and keep the changes to minimum.

Reply:
This is a product decision. Should not be based on theoretical realloc cost.

Reply:
@ashtul please remove the realloc